sklearn/preprocessing/tests/test_target_encoder.py

```diff
@@ -0,0 +1,304 @@
+import numpy as np
+from numpy.testing import assert_allclose
+from numpy.testing import assert_array_equal
+import pytest
+
+from sklearn.preprocessing import TargetEncoder, LabelEncoder
+from sklearn.model_selection import KFold
+
+
+def _encode_target(X_int, y_int, n_categories, smooth):
+    cur_encodings = np.zeros(n_categories, dtype=np.float64)
+    y_mean = np.mean(y_int)
+
+    if smooth == "auto":
+        y_variance = np.var(y_int)
+        for c in range(n_categories):
+            y_subset = y_int[X_int == c]
+            n_i = y_subset.shape[0]
+
+            if n_i == 0:
+                cur_encodings[c] = y_mean
+                continue
+
+            y_subset_variance = np.var(y_subset)
+            m = y_subset_variance / y_variance
+            lambda_ = n_i / (n_i + m)
+
+            cur_encodings[c] = lambda_ * np.mean(y_subset) + (1 - lambda_) * y_mean
+        return cur_encodings
+    else:  # float
+        for c in range(n_categories):
+            y_subset = y_int[X_int == c]
+            current_sum = np.sum(y_subset) + y_mean * smooth
+            current_cnt = y_subset.shape[0] + smooth
+            cur_encodings[c] = current_sum / current_cnt
+        return cur_encodings
+
+
+@pytest.mark.parametrize(
+    "categories, unknown_value",
+    [
+        ([np.array([0, 1, 2], dtype=np.int64)], 4),
+        ([np.array([1.0, 3.0, np.nan], dtype=np.float64)], 6.0),
+        ([np.array(["cat", "dog", "snake"], dtype=object)], "bear"),
+        ("auto", 3),
+    ],
+)
+@pytest.mark.parametrize("seed", range(2))
+@pytest.mark.parametrize("smooth", [5.0, "auto"])
+@pytest.mark.parametrize("target_type", ["binary", "continuous"])
+def test_encoding(categories, unknown_value, seed, smooth, target_type):
+    """Check encoding for binary and continuous targets."""
+
+    X_int = np.array([[0] * 20 + [1] * 30 + [2] * 40], dtype=np.int64).T
+    n_categories = 3
+    n_samples = X_int.shape[0]
+
+    if isinstance(categories, str) and categories == "auto":
+        X_input = X_int
+    else:
+        X_input = categories[0][X_int]
+
+    rng = np.random.RandomState(seed)
+
+    if target_type == "binary":
+        y_int = rng.randint(low=0, high=2, size=n_samples)
+        target_names = np.asarray(["cat", "dog"], dtype=object)
+        y_input = target_names[y_int]
+    else:  # target_type == continuous
+        y_int = rng.uniform(low=-10, high=20, size=n_samples)
+        y_input = y_int
+
+    # compute encodings for all data to validate `transform`
```

Comment:
I guess I would do that further down close to where we actually validate transform, but just a nit...